craft_application/models/project.py

```diff
@@ -64,3 +78,7 @@ def effective_base(self) -> Any:  # noqa: ANN401 app specific classes can improv
         if self.base is not None:
             return self.base
         raise RuntimeError("Could not determine effective base")
+
+    @abc.abstractmethod
+    def get_build_plan(self) -> List[BuildInfo]:
+        """Obtain the list of architectures and bases from the project file."""
```

Comment:
Applications must define their own build plan resolver because platforms definition is not uniform. This can be changed later if a common format is adopted.